modules/features2d/src/orb.cpp

```diff
@@ -650,6 +650,19 @@ static inline float getScale(int level, int firstLevel, double scaleFactor)
     return (float)std::pow(scaleFactor, (double)(level - firstLevel));
 }
 
+static inline int getLevel(KeyPoint &kp)
+{
+    int level = kp.octave;
+    CV_Assert(0 <= level);
+    //SIFT encodes layer and octave together
+    //See issue #10555
+    if (level >= 255)
+    {
+        level &= 255;
+        if (level >= 128) level |= -128;
```

Comment:
I believe, this line makes the level index negative when it's &gt; 128. Is that intended behaviour? We have tests for detectors and descriptors that check how well they preserve scale and rotation variance. Could you add such tests for SIFT + ORB combination?

Reply:
@vpisarev Yes, this is intended behavior. Please check my other comments.
As for the any potential tests, unfortunately they will fail, because of #10561, and we cannot add them now!
However, I have carefully debugged this patch, and it is sound and correct.

Reply:
@vpisarev The code has now some additional comments to explain the intended behavior.
Further, [this repo](https://github.com/Sahloul/opencv_contrib/commits/bug_fix/ORB) contains invaraince tests, and should be merge with this PR.

BTW: While the buildbot loaded my opencv_contrib repo, it did not execute its [perf_xfeatures2d](http://pullrequest.opencv.org/buildbot/builders/precommit_linux64/builds/14690) tests...
Although I have locally confirmed that tests are passing, is your development cycle really alright?

Reply:
@Sahloul You should create PR into opencv_contrib to run contrib tests.

Reply:
@alalek All done!